core/src/commonMain/kotlin/SharedRepeatableAction.kt

```diff
@@ -0,0 +1,79 @@
+package com.juul.kable
+
+import kotlinx.atomicfu.locks.reentrantLock
+import kotlinx.atomicfu.locks.withLock
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Deferred
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.async
+import kotlinx.coroutines.cancelAndJoin
+import kotlinx.coroutines.job
+import kotlin.coroutines.CoroutineContext
+
+/**
+ * A mechanism for launching and awaiting a shared action ([Deferred]) repeatedly.
+ *
+ * The [action] is started by calling [await]. Subsequent calls to [await] will return the same
+ * (i.e. shared) [action] until cancelled or failure occurs.
```

Comment:
Just checking that I'm reading this correctly: if the action completes successfully, subsequent calls to `await` will return the cached result.

Reply:
Correct. That is intentional. This is primarily for connection handling.
If connect `action` completed successfully, we'll still return the same `Deferred` to await on. Since it will already be done, the `await` will return immediately.
Since coroutines may have been launched from the `action`, those could still be running, if they fail at any point, then it resets the "repeatable action".